src/ray/raylet/node_manager.cc

```diff
@@ -341,22 +341,40 @@ void NodeManager::Heartbeat() {
   auto heartbeat_data = std::make_shared<HeartbeatTableData>();
   SchedulingResources &local_resources = cluster_resource_map_[self_node_id_];
   heartbeat_data->set_client_id(self_node_id_.Binary());
+
   // TODO(atumanov): modify the heartbeat table protocol to use the ResourceSet directly.
   // TODO(atumanov): implement a ResourceSet const_iterator.
-  for (const auto &resource_pair :
-       local_resources.GetAvailableResources().GetResourceMap()) {
-    heartbeat_data->add_resources_available_label(resource_pair.first);
-    heartbeat_data->add_resources_available_capacity(resource_pair.second);
+  if (!last_heartbeat_resources_.GetAvailableResources().IsEqual(
+          local_resources.GetAvailableResources())) {
+    for (const auto &resource_pair :
+         local_resources.GetAvailableResources().GetResourceMap()) {
+      heartbeat_data->add_resources_available_label(resource_pair.first);
+      heartbeat_data->add_resources_available_capacity(resource_pair.second);
+    }
+    last_heartbeat_resources_.SetAvailableResources(
+        ResourceSet(local_resources.GetAvailableResources()));
   }
-  for (const auto &resource_pair : local_resources.GetTotalResources().GetResourceMap()) {
-    heartbeat_data->add_resources_total_label(resource_pair.first);
-    heartbeat_data->add_resources_total_capacity(resource_pair.second);
+  if (!last_heartbeat_resources_.GetTotalResources().IsEqual(
```

Comment:
Let's say every other resources are updated but this. In this case, isn't GCS going to update the false information? It is because 
```
if (heartbeat_data.should_global_gc() ||
      heartbeat_data.resources_available_label_size() > 0 ||
      heartbeat_data.resources_total_label_size() > 0 ||
      heartbeat_data.resource_load_label_size() > 0) {
    heartbeat_buffer_[node_id] = heartbeat_data;
```
this code block will still update the `heartbeat_data` in `heartbeat_buffer_`, but `heartbeat_data` says this resources will be empty? Did I miss something here?

Reply:
gcs will always lose previous data after broadcasting heartbeat batch done. (by invoking `hearbeat_buffer_.clear()`)
this `if` block check if there's gc_flag/resource update in heartbeat. if there's any, add it in buffer then broadcast it. else do nothing.

Reply:
I see. That explains every question I have. Thanks for the explanation!